modules/mauth-authenticator-apachehttp/src/main/java/com/mdsol/mauth/apache/HttpClientPublicKeyProvider.java

```diff
@@ -96,18 +96,35 @@ private String getRequestUrlPath(UUID appUUID) {
   }
 
   private class PublicKeyResponseHandler implements ResponseHandler<String> {
+    private static final String MAX_AGE = "max-age";
+    private static final String PUBLIC_KEY_STR = "public_key_str";
 
     @Override
     public String handleResponse(HttpResponse response) throws IOException {
       if (response.getStatusLine().getStatusCode() == HttpStatus.SC_OK) {
+        if (ttl == null) {
+          ttl = getMaxage(response).orElse(configuration.getTimeToLive());
+        }
+
         HttpEntity entity = response.getEntity();
         String responseAsString = EntityUtils.toString(entity, StandardCharsets.UTF_8);
         ObjectMapper mapper = new ObjectMapper();
-        return mapper.readTree(responseAsString).findValue("public_key_str").asText();
+        return mapper.readTree(responseAsString).findValue(PUBLIC_KEY_STR).asText();
       } else {
         throw new HttpClientPublicKeyProviderException("Invalid response code returned by server: "
-            + response.getStatusLine().getStatusCode());
+          + response.getStatusLine().getStatusCode());
+      }
+    }
+
+    public Optional<Long> getMaxage(HttpResponse response) {
+      if (response.getFirstHeader(HttpHeaders.CACHE_CONTROL) != null) {
+        HeaderElement[] elements = response.getFirstHeader(HttpHeaders.CACHE_CONTROL).getElements();
+        return Arrays.stream(elements)
+          .filter(e -> e.getName().equals(MAX_AGE))
+          .findFirst()
+          .map(e -> Long.parseLong(e.getValue()));
```

Comment:
Response cache header max-age value can only be a numeric value in seconds so there is no way this would fail unless the cache-headers are tampered with. So should we allow public key storage of response with tampered headers?

Reply:
Considering the tampered headers scenario makes sense to throw an exception. So the request to the service consuming this library will return 401.
Do you think there is a possibility of the Mauth service returning an invalid value here?, because in that case, the consequence will be all consumers of this library returning 401 for all requests, which could be avoided by using configuration.getTimeToLive()

Reply:
Max-age can only be numeric, and it's impossible for mauth to have the wrong value since the Ruby framework is responsible for setting up the cache control headers (and making sure that the 'expires' value is numeric). Also there is a cucumber test in mauth to make sure that the value is valid.

Reply:
Looks like we're covered then, merging EOD unless further objections.

Reply:
Looks good to me, thanks for checking